graph_weather/models/gencast/graph/graph_builder.py

```diff
@@ -281,16 +286,59 @@ def _init_mesh2grid_graph(self):
         return m2g_graph
 
     def _init_khop_mesh_graph(self):
-        """Build k-hop Mesh graph."""
+        """Build k-hop Mesh graph.
 
-        transform = TwoHop()
-        khop_mesh_graph = self.mesh_graph
-        for _ in range(self.num_hops):
-            khop_mesh_graph = transform(khop_mesh_graph)
+        This implementation constructs the sparse adjacency matrix associated with the mesh graph
+        and computes its powers in a sparse manner. It should be more memory-efficient than the
+        original PyG implementation because it does not need to materialize all the edges.
+        """
 
+        # PyG version:
+        # transform = TwoHop()
+        # khop_mesh_graph = self.mesh_graph
+        # for _ in range(self.num_hops):
+        #    khop_mesh_graph = transform(khop_mesh_graph)
+
+        # build the sparse adjacency matrix
+        edge_index = self.mesh_graph.edge_index
+        adj = torch.sparse_coo_tensor(
+            edge_index,
+            values=torch.ones_like(edge_index[0], dtype=torch.float32),
+            size=(self._num_mesh_nodes, self._num_mesh_nodes),
+        ).to(
+            "cpu"
```

Comment:
Does this need to be on the CPU? Seems like it shouldn't be hardcoded at least, even if it does seemingly make it more memory efficient

Reply:
Added a "khop_device" option for selecting the device!